WIP SIP-23

adriaanm opened the pull request:
WIP PR to centralize discussion, accept improvements/tests/polish/docs/...

Everyone is welcome to look for the checkboxes/TODO items here and in the discussion below and take them on! If you're not sure where/how to start, just ask!
# Testing
- [ ] syntax (infix notation, backquotes, quasiquotes, string interpolation?,...)
- [ ] subtyping
- [ ] implicits
- [ ] dependent method types
- [ ] type inference: not inferred for signatures, inferred when <: Singleton
# Spec/docs
- [ ] include negated literal types in InfixType BNF (like prefixExpr)
- [ ] update syntax overview in spec
- [ ] spec the subtyping relationship between `ConstantType` / `LiteralType` and `scala.Singleton`.
- [ ] include SIP doc under spec/
# Type inference / type checking
- [x] When inferring signatures, treat literal types like singleton types (i.e., don't infer them -- widen instead)
- [x] treat singleton type bound on type param same as Singleton bound (stopWidening when bounds exists _.isStable)
- [x] include fix for https://issues.scala-lang.org/browse/SI-8564 (idm.deconst)
- [x] lub/glb
- [x] instanceOf checks: `x.isInstanceOf[3]` should expand to `x == 3`, and `x.asInstanceOf[3]` should be `if (x != 3) throw new ClassCastException(..)`
# Corner cases
- [ ] `val t: 1 = t`  (NOTE: t has value 0...)
# Design space
- [x] Support negative literals
- [ ] Support Symbol literals --> challenge: desugars into Symbol.apply("symbol")
- [ ] Support class literals? (probably not)
- [x] Model `()` as `LiteralType`? (I say 'no': () is not a syntactic literal, it's also too deeply ingrained to incorporate here)
# Feature interaction
- [ ] Predicate on language flag (can't do this in parser, though -- use -Xsip:23 instead?)
- [ ] Update quasiquotes 
- [ ] Inconsistent representation of parse trees in different contexts
- [ ] scalap

adriaanm committed:
WIP SI-5103 avoid widening TypeVar <: Singleton

TODO: deriving this from type param bounds is not the best way,
should consider all constraints that we encounter

scala> def stable[T <: 1](x: T): T = x
stable: [T <: 1](x: T)T

scala> stable(1)
res0: 1 = 1

scala> stable(2)
<console>:9: error: inferred type arguments [2] do not conform to method stable's type parameter bounds [T <: 1]
              stable(2)
              ^
<console>:9: error: type mismatch;
 found   : Int(2)
 required: T
              stable(2)
                     ^

scala> def stable[T <: Singleton](x: T): T = x
stable: [T <: Singleton](x: T)T

scala> val x: Int = 2
x: Int = 2

scala> val y: x.type = x
y: x.type = 2

scala> stable(y)
res3: x.type = 2

adriaanm committed:
bump module versions in eclipse projects

adriaanm committed:
[WIP] SIP-23 Parse `Literal` as `SimpleType`

This requires -Xexperimental in 2.11.x.

Update spec.

TODO:
 - test!
 - include SIP as spec addendum
 - don't predicate on -Xexperimental -- introduce -Xsip:23?

Notes:
 - `()` is not a literal, neither are `Symbol`s

adriaanm committed:
[WIP] SIP-23 LiteralType: 42 and friends

`LiteralType` is a `SingleType` with a literal for its path.
With an example: `1` is to `p.type`, as `1` is to `p`.

`lit` is the most precise type for an expression known to evaluate to `lit`.
Note that we cannot constant fold the expression to `lit`, since it may have a side effect.

`ConstantType(c)` is the type of an expression that can be constant folded to `c`.

When an expressions should not be constant folded, the corresponding `LiteralType`
is created: `ConstantType(c).deconst == LiteralType(c)`
(this is where all `LiteralType`s originate, except for unpickling).

`typedLiteral` always infers a `ConstantType`, which is subsequently
`deconst`ed or `widen`ed where necessary (see `widenIfNecessary`).
In short, only `final val`s may have `ConstantType`s, and
`widen`ing is required to ensure that a path is stable and accessible (roughly speaking).

When checking subtyping, this invariant is useful: `LiteralType(c).widen == c.tpe`.

TODO:
  - test! (subtyping, type inference, type tests, patmat, dependent method types)
  - fully predicate implementation on -Xsip:23
  - is it safe to reduce LITERAL in pickler?

adriaanm committed:
[WIP] SIP-23 equality test for singleton type <:< AnyVal

test case:

```
val y: 5 = 5
def g(x: Int) = x match {
  case _: y.type => 0
}
```

adriaanm commented:
```
scala/ (sip23 $=) $ qs -Xexperimental
Welcome to Scala version 2.11.5-20141028-095954-500f1ab300 (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0).
Type in expressions to have them evaluated.
Type :help for more information.

scala> val x = 1
x: 1 = 1

scala> val y = 2
y: 2 = 2

scala> (x + 1) match { case _ : y.type => 3 }
res1: 3 = 3

scala> def foo = { println(x); y }
foo: Int

scala> def stable[T <: Singleton](x: T): T = x
stable: [T <: Singleton](x: T)T

scala> stable(y)
res3: 2 = 2

scala> identity(y)
res4: Int = 2
```

adriaanm commented:
```
scala> x.isInstanceOf[y.type]
res6: Boolean = false

scala> x.isInstanceOf[1]
res7: Boolean = true
```

adriaanm committed:
[WIP] SIP-23 fix doc comment in parsers

adriaanm commented on src/compiler/scala/tools/nsc/typechecker/Implicits.scala:
```diff
@@ -1160,6 +1160,7 @@ trait Implicits {
           // necessary only to compile typetags used inside the Universe cake
           case ThisType(thisSym) =>
             gen.mkAttributedThis(thisSym)
+          // TODO SIP-23: TypeTag for LiteralType
```
- [ ] This could serve as the vehicle for summoning the type's inhabitant.

xeno-by replied:
Why not have a macro for this?

adriaanm commented on src/reflect/scala/reflect/api/Types.scala:
```diff
@@ -565,6 +570,36 @@ trait Types {
   }
 
+  /** The `LiteralType` type represent a singleton type with a literal for its path.
+   *  Examples: "a".type or 42.type, which may be abbreviated to "a" or 42.
```
- [ ] TODO: update to syntax without `.type` suffix

adriaanm commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -5127,13 +5127,17 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
       }
 
       def typedSingletonTypeTree(tree: SingletonTypeTree) = {
+        // SIP-23: don't require AnyRef for 1.type etc
```
- [ ] TODO: update to syntax without `.type` suffix

folone commented:
This looks amazing, @adriaanm! I'm playing with it at the moment. Got a few questions:
1. Is the plan to allow `.type` syntax only on non-literals? Or will we have an optional `.type` on literals as well?
   
   ``` scala
   scala> val test: 10 = 10
   test: 10 = 10
   
   scala> val test: 10.type = 10
   <console>:1: error: ';' expected but '.' found.
          val test: 10.type = 10
                      ^
   ```
2. The example I was struggling with the most is now working (yay!):
   
   ``` scala
     scala> def ok: 7 = {
        | println("yay")
        | 7
        | }
   ok: 7
   
   scala> ok
   yay
   res10: 7 = 7
   ```
3. This is just awesome:
   
   ``` scala
   scala> val x = 1
   x: 1 = 1
   
   scala> val y: x.type = 1
   y: x.type = 1
   
   scala> implicitly[x.type =:= y.type]
   res16: =:=[x.type,y.type] = <function1>
   
   scala> implicitly[1 =:= y.type]
   res17: =:=[1,y.type] = <function1>
   ```
4. The issue that you've fixed in [SI-8564](https://issues.scala-lang.org/browse/SI-8564) does not work in this branch though:
   
   ``` scala
   scala> def noisyIdentity(x: Any): x.type = {
        |     println("got " + x)
        |     x
        |   }
   noisyIdentity: (x: Any)x.type
   
   scala> noisyIdentity("test")
   res11: "test" = test
   
   scala> noisyIdentity(new java.lang.String("test"))
   got test
   res12: String = test
   ```
5. Parsing unary minus is not working:
   
   ``` scala
   scala> val test: -4 = -4
   <console>:1: error: ';' expected but integer literal found.
          val test: -4 = -4
                     ^
   
   scala> val test = -4
   test: -4 = -4
   ```
6. Some more random rough edges I could find:
   
   ``` scala
   scala> val t: 1 = t
   <console>:9: warning: value t does nothing other than call itself recursively
          val t: 1 = t
                     ^
   t: 1 = 0 // jvm default value?
   ```
   
   ``` scala
   scala> trait Succ[T] {
        |     type Out
        |     def apply(x: T): Out
        |   }
   defined trait Succ
   
   scala> implicit object One extends Succ[1] {
        |     type Out = 2
        | def apply(x: 1) = 2
        | }
   defined object One
   
   scala> def f[T](x: T)(implicit succ: Succ[T]) = succ(x)
   f: [T](x: T)(implicit succ: Succ[T])succ.Out
   
   scala> implicitly[Succ[1]]
   res13: Succ[1] = One$@735d537e
   
   scala> f(1)
   <console>:12: error: could not find implicit value for parameter succ: Succ[Int]
                 f(1)
                  ^
   ```

Hope this is helpful. I'm super excited to see this moving forward! :100: :+1:

adriaanm commented:
Thanks for the feedback -- very helpful! I have to work on my devoxx talk for the next couple of days, but will get back to coding on this soon. In the mean time, PRs accepted on my sip23 branch ;-)

folone commented:
I'll definitely spend quite some time with it!

adriaanm commented:
Regarding the last rough edge, it's probably related to lack of type equality between ConstantType and LiteralType. Still mulling that one over.

retronym commented:
Type inference for vals becomes more specific, is this intended?

```
scala> val y = 1
y: 1 = 1

scala> val y = if (true) 1 else 1
y: 1 = 1

scala> val y = if ("".isEmpty) 1 else 1
y: 1 = 1
```

adriaanm commented:
It is, though it's up for debate. Under the hood, a `ConstantType` now `deconst`s to a `LiteralType`, which only then widens to the underlying type. Before, `deconst` would also `widen` (because there was no more precise type that wasn't a ConstantType).

retronym commented:
What's the plan for class literals? The corresponding `LiteralType` can be inferred, but not expressed in syntax:

```
scala> val y = classOf[String]
y: classOf[java.lang.String] = class java.lang.String
```

Again, I think tend to think we should not infer these types unless guided by a bound of `Singleton`.

adriaanm commented:
I was planning to only support [syntactic literals](https://github.com/scala/scala/blob/2.11.x/spec/01-lexical-syntax.md#literals). Right now, Symbols are not supported.

retronym commented:
- [ ] One small thing that is missing at the moment is a spec for the subtyping relationship between `ConstantType` / `LiteralType` and `scala.Singleton`.

adriaanm commented:
The problem with Symbols is similar to classOf, they are method calls underneath:

```
scala> val x: 's = 's
<console>:7: error: stable identifier required, but scala.Symbol.apply("s") found.
       val x: 's = 's
```

adriaanm commented:
> Again, I think tend to think we should not infer these types unless guided by a bound of Singleton.

Do you have any snags in mind if we do?

retronym commented:
Just the general problem of accidentally inferring too specific a type for some public API that ties your hands down the track. IIUC, that's why we deconst in type inference today, rather than just to ensure side effects aren't folded away.

I'd expect var and val infer the same type.

```
scala> val x = if ("".isEmpty) 1 else 1
x: 1 = 1

scala> var x = if ("".isEmpty) 1 else 1
x: Int = 1
```

adriaanm commented:
As a compromise, we could widen for non-local vals. I think it's useful to preserve precise types locally.

retronym commented:
Different rules for different contexts doesn't sound ideal to me. But we can give it some thought.

adriaanm commented:
It feels desirable to me. `widenIfNecessary` already implements different rules for vars (`shouldWiden`) and (final) vals.

adriaanm commented:
@odersky, what do you think? Is inferring `x: 1` for `val x = 1` desirable?

retronym commented:
While we ponder that one, here's a curious wrinkle:

```
scala> List[1](1)
<console>:8: error: type mismatch;
 found   : Array[1]
 required: Array[Int]
Note: 1 <: Int, but class Array is invariant in type T.
You may wish to investigate a wildcard type such as `_ <: Int`. (SLS 3.2.10)
              List[1](1)
                     ^
```

adriaanm commented:
Nice... covariant array ftw

adriaanm commented:
Probably a widen in varargs code?

retronym committed:
SIP-23 Account for literal types in varargs translation

Before:

	sandbox/test.scala:2: error: type mismatch;
	 found   : Array[1]
	 required: Array[Int]

retronym commented:
https://github.com/adriaanm/scala/pull/13

adriaanm commented:
Thanks!

retronym commented:
Here's a failing test case with Java varargs.

https://github.com/retronym/scala/compare/topic/sip23-java-varargs?expand=1

In `foundReqMessage`, we find:

```
found = Array[<LiteralType(1)>]
req = Array[<RefinedType(LiteralType(1), Object)>]
```

The latter type is incorrectly printed as `Array[1]`.

retronym commented:
This works as expected, but is the sort of thing we'll have to add test cases for:

```
scala> class C { def foo(x: 1) = 0; def foo(x: Int) = 0 }
<console>:33: error: double definition:
def foo(x: 1): Int at line 33 and
def foo(x: Int): Int at line 33
have same type after erasure: (x: Int)Int
       class C { def foo(x: 1) = 0; def foo(x: Int) = 0 }
                                        ^
```

retronym commented:
This seems inconsistent:

```
scala> class C[X <: Singleton](x: X)
defined class C

scala> new C(1)
res10: C[1] = C@7cd3860
```

vs

```
scala> class C[X <: 1](x: X)
defined class C

scala> new C(1)
<console>:35: error: inferred type arguments [Int] do not conform to class C's type parameter bounds [X <: 1]
              new C(1)
              ^
<console>:35: error: type mismatch;
 found   : Int(1)
 required: X
              new C(1)
                    ^

```

retronym commented:
The spec for erasure (in "03. Types") currently says:

```
- The erasure of a singleton type `$p$.type` is the
  erasure of the type of $p$.
```

`ConstantType`-s aren't directly mentioned. Today, they are left intact by erasure, but later widened in the backend. I'm planning to change things to so they are widened during erasure; this is to fix:

```
scala> object O { override final val toString = "" }
java.lang.ClassFormatError: Duplicate method name&signature in class file O$
```

We should review that area of the spec and implementation holistically for that change and SIP-23.

retronym commented on src/reflect/scala/reflect/internal/Types.scala:
```diff
@@ -2781,18 +2807,24 @@ trait Types
+        }
+        else new TypeConstraint
+
+      if (bounds.hi.contains(SingletonClass)) constr.stopWidening()
```
Nice!

retronym commented:
Not right yet, but when changing the pickler we also need to update scalap

adriaanm commented:
> Different rules for different contexts doesn't sound ideal to me. But we can give it some thought.

I ran it by @odersky over lunch and he agrees. We should treat literal types like singleton types, like you said.

retronym commented:
- [ ] Update quasiquotes 

```
scala> type T = 1
defined type alias T

scala> val tq"1" = typeOf[T].dealias
scala.MatchError: 1 (of class scala.reflect.internal.Types$UniqueLiteralType)
  ... 38 elided

scala> tq"1"
res3: $r.intp.global.SingletonTypeTree = 1.type

scala> q"val x: 1 = 1"
res4: $r.intp.global.ValDef = val x: 1.type = 1
```

adriaanm committed:
SI-8564

folone committed:
Add some basic smoke tests for SIP-23.

folone commented:
Added some quick things in #14

retronym commented:
I wonder if we ought to introduce a new magic type, `Literal`, ala `Singleton`.

See:
https://twitter.com/extempore2/status/527241990363049984
https://twitter.com/extempore2/status/527242133124571136

clhodapp commented:
Ha: I had a similar idea at almost the same time and performed a test: https://gist.github.com/clhodapp/7037e4058698e3cea490. It seems that Singleton already kind of acts this way, though I don't feel that the fact that implicits for Singleton will match literals but not stable identifiers is currently sufficiently well-documented to actually depend on. Perhaps this just needs to be documented as the behavior and tests added to prevent it from changing? Or maybe some corner cases exist that  do make introducing a Literal type necessary?

retronym commented:
Yep, `Literal` would be a subtype of `Singleton`. It would close the `sToA(x)` escape hatch you identified.

folone committed:
Add LiteralType to JavaUniverseForce, backward, and forward bincompat.

adriaanm commented:
The only difference I can think of between `Literal` and `Singleton` is that a literal type's value is known syntactically, but when you start abstracting over it (which is what `Literal` is for), that is lost. So, why not just use `Singleton` for both? A literal type is a path-dependent type, where the path is a literal. Both are singleton types.

My example above already works:

```
scala> def stable[T <: Singleton](x: T): T = x
stable: [T <: Singleton](x: T)T

scala> stable(y)
res3: 2 = 2
```

retronym commented:
You're right, let's leave out `Literal` unless a compelling use case comes forth.

adriaanm committed:
[WIP] SIP-23: accept negated literal in infixTypeRest

e.g., val x : -1 => -1 = ???

densh commented on src/reflect/scala/reflect/internal/Types.scala:
```diff
@@ -1263,6 +1265,28 @@ trait Types
+   * This type is created by `deconst`ing the corresponding `ConstantType` (see typedLiteral and widenIfNecessary);
+   * it `widen`s to the type of `value`.
+   *
+   * NOTE: `()` is not a literal, and neither are `Symbol`s (`'a`)
```
Why not? It looks like there is some [interest in symbol singleton types](https://github.com/typelevel/scala/issues/51).

densh replied:
It would also be really nice if we could just make Unit an alias to `()` singleton type.

adriaanm replied:
the remark about symbols is a mistake

adriaanm replied:
unit it special enough that I think we should leave it as is (I think of it as Tuple0).

densh replied:
Why not alias the other way around for sake of consistency then? (e.g. in pseudocode `type () = Unit`)

It might be a bit problematic for meta if unit is special wrt literal types.

adriaanm replied:
How so? The rule is simple: the syntactic category of literals is included in LiteralType. () is not a Literal in the BNF grammar. Changing these things around is costly. We'll need a very convincing reason to do so.

densh replied:
Thanks for pointing this out. We assumed that `()` is a literal syntax-wise and typed it accordingly in our strongly typed tree api which one might consider to be a natural way to model Scala literals (considering the fact that `()` is still called unit literal.)

If support for unit literal is costly we might have to re-do this part of our trees. But I'm not sure what exactly are the costs you are talking about.  Grammar just needs another case around https://github.com/adriaanm/scala/pull/12/files#diff-c067e953d98d0b8fb993aeae8bdb0bd4R26 that will allow `"()"` as a valid type. I don't think that this will cause any ambiguities apart from `() => T` which can be resolved in favour of current interpretation as `Function0[T]`. Tree produced for `()` type can be just `_root_.scala.Unit` if `LiteralType(())` is problematic to implement.

I don't mind contributing required changes to the parser/grammar.

densh replied:
/cc @xeno-by

adriaanm replied:
Sorry for being vague. I am thinking about the hidden assumptions in the code base regarding ()/Unit's current encoding. Parsing would not be hard, I agree. I'm more concerned about all the special casing it's received both in the batch and presentation compilers.

densh replied:
The syntax can just alias `_root_.scala.Unit` (i.e. desugared in parser). This way it won't be distinguishable from regular unit after parsing. Wdyt?

adriaanm replied:
My concern is with the type representation, right now `()` has type `TypeRef(scala, UnitClass, Nil)`

retronym commented:
We should also consider whether this language change can help us get closer to writing:

``` scala
def identity[A](a: A): a.type = a
```

densh commented:
It would absolutely amazing if this also happens to fix [SI-5103](https://issues.scala-lang.org/browse/SI-5103).

adriaanm commented:
IT depends on the use case, but the simple one works (see `def stable` above).

adriaanm committed:
WIP: quickfix for nopos in parser -- proper fix?

adriaanm committed:
widenIfNecessary

adriaanm committed:
isSuitableLiteralType

adriaanm committed:
widen when checking for typeref

adriaanm committed:
simplify subtyping of SingletonType?

adriaanm committed:
sip23 condition

this bootstraps

adriaanm commented:
Here's a limited (but working) example of how you can simplify an HMap implementation:

``` scala
import scala.language.dynamics


class Rec[+KV <: (_ <: Singleton) ~> _] extends Dynamic {
  def selectDynamic(key: Any)(implicit lu: Lookup[KV, key.type]): lu.T = ???
}

trait Lookup[-KV <: (_ <: Singleton) ~> _, K] { type T }
object Lookup {
  implicit def luK[K <: Singleton, V] = new Lookup[K ~> V, K] { type T = V}
}

trait ~>[K <: Singleton, +V]
trait Types {
  type R1String2Int = Rec[("l1" ~> String) with ("l2" ~> Int)]
  type R1Any2Int = Rec[("l1" ~> Any) with ("l2" ~> Int)]
  type R2Int1String = Rec[("l2" ~> Int) with ("l1" ~> String)]
  type R1String = Rec["l1" ~> String]
  type R2Int = Rec["l2" ~> Int]
}

object Subtyping extends Types {
  // width
  val s1: R1String = (??? : R1String2Int)
  val s2: R2Int = (??? : R1String2Int)

  // depth
  val s1Any2: R1Any2Int = (??? : R1String2Int)

  // permute
  val s21: R2Int1String = (??? : R1String2Int)
}

object Projection extends Types {
  val sel1: String = (??? : R1String2Int).l1 // doesn't work for l2 :-( // selectDynamic("l1") 
}
```

adriaanm commented:
Hmm, weird. I won't have time until next week to work on this. I guess I un-fixed something then...

folone commented:
Just found another weirdness:

``` scala
$ ./build/pack/bin/scala -Xexperimental
Welcome to Scala version 2.11.5-20141116-124950-06ec2103df (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_20).
Type in expressions to have them evaluated.
Type :help for more information.

scala> val x = 1
x: 1 = 1

scala> val x: 2 = 1
<console>:7: error: type mismatch;
 found   : Int(1)
 required: 2
       val x:2 = 1
                 ^

scala> val x: Int = 1
x: Int = 1

scala> x.asInstanceOf[1]
res0: 1 = 1

scala> x.asInstanceOf[2]
res1: 2 = 1

scala> x.asInstanceOf[3]
res2: 3 = 1
```

adriaanm commented:
Good catch :) can you add an example under Corner Cases in our todo list? Should have some time this week to work on this

retronym commented:
I guess `x.isInstanceOf[3]` should expand to `x == 3`, and `x.asInstanceOf[3]` should be `if (x != 3) throw new ClassCastException(..)`. This is analagous to the translation of `x.isInstanceOf[outer.T]`, which checks if `x.outer$ eq outer`.

adriaanm commented:
Yep -- this translation is already implemented for pattern matches. I forgot to duplicate the fix where casts are emitted.

folone commented:
@adriaanm I'm not allowed to edit the description :)

adriaanm commented:
Ah, that's unfortunate. I'll add it then.

xuwei-k commented:
> and x.asInstanceOf[3] should be if (x != 3) throw new ClassCastException(..)

Is this a bug? or specification?

``` scala
Welcome to Scala version 2.11.4 (Java HotSpot(TM) 64-Bit Server VM, Java 1.7.0_67).
Type in expressions to have them evaluated.
Type :help for more information.

scala> final val a = "a"
a: String("a") = a

scala> "b".asInstanceOf[a.type]
res0: String = a
```

adriaanm commented:
Thanks, yep, that's a bug! The same one as `1.isInstanceOf[2]`, in fact -- will be fixed in same way.

adriaanm committed:
sip23: numericLub

adriaanm commented:
Yay, FYI, @densh, it bootstraps. (Suboptimal solution)

adriaanm committed:
asInstanceOf for singleton types

```
scala> val x = "a"
x: "a" = a

scala> x.asInstanceOf["a"]
res1: "a" = a

scala> "1".asInstanceOf["2"]
x asInstanceOf Throwable
java.lang.ClassCastException
  ... 33 elided

scala> "1".asInstanceOf["1"]
res3: "1" = 1

scala> 1.asInstanceOf[1]
res4: 1 = 1

scala> 1.asInstanceOf[2]
x asInstanceOf Throwable
java.lang.ClassCastException
  ... 33 elided

```

densh commented:
Yay, singleton types get inferred now for the usecase I've had:

``` scala
scala> class Region
defined class Region

scala> class Ref[R <: Region with Singleton]
defined class Ref

scala> val r = new Region
r: Region = Region@66cd51c3

scala> val p = new Ref[r.type]
p: Ref[r.type] = Ref@2d363fb3

scala> def f[T <: Region with Singleton](x: Ref[T]) = x
f: [T <: Region with Singleton](x: Ref[T])Ref[T]

scala> f(p)
res0: Ref[r.type] = Ref@2d363fb3
```

Looking forward to this. :+1:

adriaanm commented:
Cool! Could you write down a spec for the cases in which you'd expect a singleton type to be inferred? Right now, this happens for stable values that do not refer to a hidden variable. The main difference in behavior was that `deconst` used to `widen`, now it converts a `ConstantType` to a `LiteralType`, so we infer much stricter types for non-final vals (which used to do a `deconst.widen`, but now we only do `deconst`).

soc commented:
What's the perspective here regarding union types?

If we have something like `if (bool) 1 else 2` would the return type be `1 | 2`?

adriaanm commented:
Likely, but union types are still too far in the future to know for sure how they'll interact with this feature. They are primarily about class types -- you could already experiment in dotty and see what p.type | q.type means, where p and q refer to strings...

adriaanm commented:
(Will reopen when I get a chance to work on this again!)
